### DISCUSSION TITLE
Add setting or mode to automatically expand arrays

### DISCUSSION THREAD

**Discussion by wdavis** [spatie/ray#20] [Ideas]
Just purchased Ray. It's great!

I've been spending quite a bit of time expanding arrays and objects by hand while using Ray.

**It would be nice to be able to auto expand/collapse items as a mode or even a button per entry.**

As an alternative I tried:

```php
ray(json_encode($array, JSON_PRETTY_PRINT));
```

But that doesn't quite format in a way that is very helpful -- would need to be wrapped in a `<pre></pre>` or something.

Any recommendations?

**Comment by ejunker** [comment]
Good idea. I looked into it a bit and as far as I can tell Ray uses symfony/var-dumper. I found a post on Laracasts asking about expanding dumps by default. https://laracasts.com/discuss/channels/general-discussion/expanding-dd-vardumper-by-default

**Reply by wdavis** [reply]
Modifying var-dumper may be the wrong direction.

Instead, with the custom macros it looks like you can do something like this. Here's a rough POC:

```php
Ray::macro('myCustomFunction', function($something) {
    ray()->sendCustom("<pre>{$something}</pre>");
});
```

Then call it:

```php
ray()->myCustomFunction(json_encode(['one','two','three'], JSON_PRETTY_PRINT));
```

![Screen Shot 2021-01-08 at 6 10 43 PM](https://user-images.githubusercontent.com/431631/104076828-e4e80d00-51dc-11eb-9f65-a7af7b637f12.png)

With some refinement this could be a pretty neat addition.

**Reply by freekmurze** [reply]
There is an `ArgumentConverter` in the spatie/ray package that can convert any argument to a string, maybe that helps?

**Reply by vesper8** [reply]
I managed to achieve what I wanted with this macro

```
        \Spatie\Ray\Ray::macro('prettyPrint', function ($array) {
            ray()->sendCustom(sprintf('<pre>%s</pre>', json_encode($array, JSON_PRETTY_PRINT)));
        });
```

And call it like this

```
ray()->prettyPrint($nestedArray);
```
